datumaro/components/format_detection.py

```diff
@@ -180,6 +200,80 @@ def probe_text_file(
         except Exception:
             self.fail(requirement_desc_full)
 
+    @contextlib.contextmanager
+    def alternatives(self) -> Iterator[None]:
+        """
+        Returns a context manager that can be used to place a requirement that
+        is considered met if at least one of several alternative sets of
+        requirements is met.
+        To do so, use a `with` statement, with the alternative sets of
+        requirements represented as nested `with` statements using the context
+        manager returned by `alternative`:
+
+            with context.alternatives():
+                with context.alternative():
+                    # place requirements from alternative set 1 here
+                with context.alternative():
+                    # place requirements from alternative set 2 here
+                ...
+
+        The contents of all `with context.alternative()` blocks will be
+        executed, even if an alternative that is met is found early.
+
+        Requirements must not be placed directly within a
+        `with context.alternatives()` block.
+        """
+
+        self._start_requirement()
+
+        self._alternatives_context = self._AlternativesContext()
+
+        try:
+            yield
+
+            # If at least one `alternative` block succeeded,
+            # then the `alternatives` block succeeds.
+            if self._alternatives_context.had_successful_alternatives:
+                return
+
+            # If no alternatives succeeded, and none failed, then there were
+            # no alternatives at all.
+            assert self._alternatives_context.failed_alternatives, \
+                "an `alternatives` block must contain " \
+                "at least one `alternative` block"
+
+            raise FormatRequirementsUnmet(
+                self._alternatives_context.failed_alternatives)
+        finally:
+            self._alternatives_context = None
```

Comment:
I'm not sure we need to allow having more than 1 alternatives section, and having both alternatives and plain reqs.

Reply:
Why not? An `alternatives` block is just another type of requirement (just made out of other requirements). This allows detectors to combine requirements in arbitrary ways (both logical conjunctions and disjunctions).

Reply:
It is questionable if we can benefit from denormalized requirements on formats. It looks like it will be hard to implement different confidence values for different alternatives this way, but I can be wrong about this.

Reply:
What do you mean by "denormalized"?

Reply:
> What do you mean by "denormalized"?

The opposite of https://en.wikipedia.org/wiki/Disjunctive_normal_form, i.e. ability to represent conditions like "A and (B or C) ..." and not just "(A and B) or (A and C)". 

Probably, now I understood the model you're trying to implement. I want to say that maybe it is too powerful for our task and thus complicated for the user. It is an open question, which we're unlikely to answer now.

Reply:
> The opposite of https://en.wikipedia.org/wiki/Disjunctive_normal_form, i.e. ability to represent conditions like "A and (B or C) ..." and not just "(A and B) or (A and C)".

There are at least a couple reasons to allow that:

* it doesn't forcing the user to duplicate the code for A;
* it prevents evaluating A twice.

Reply:
Ok, let's keep it that way.